pkg/controller/utils/ownership.go

```diff
@@ -0,0 +1,126 @@
+	}
+
+	return nil
+}
+
+// SyncControllerReference ensures that the object passed in has a controller owner reference of the owner passed in. It then updates the object in Kube.
+// If there is already a controller owner reference set and it's set to the passed in owner, it does nothing.
+// If there is already a controller owner reference, but it's set to a different owner than the one passed in, it will set the controller owner reference to the owner that is passed in.
+// If there isn't a controller owner reference, it will add one.
+func SyncControllerReference(owner hivev1.MetaRuntimeObject, object hivev1.MetaRuntimeObject, kubeclient client.Client, scheme *runtime.Scheme, logger log.FieldLogger) error {
+	objectNamespacedName := object.GetNamespace() + "/" + object.GetName()
+	ownerNamespacedName := owner.GetNamespace() + "/" + owner.GetName()
+
+	objectGVK, err := apiutil.GVKForObject(object, scheme)
+	if err != nil {
+		logger.WithField("objectNamespacedName", objectNamespacedName).Warn("getting GVK for object")
+		return nil // Not returning error because that could stop the overall reconciliation. Just logging a warning.
+	}
+
+	ownerGVK, err := apiutil.GVKForObject(owner, scheme)
+	if err != nil {
+		logger.WithField("ownerNamespacedName", ownerNamespacedName).Warn("getting GVK for owner")
+		return nil // Not returning error because that could stop the overall reconciliation. Just logging a warning.
+	}
+
+	objectLogger := logger.WithFields(log.Fields{
+		"ownerKind":            ownerGVK.Kind,
+		"ownerNamespacedName":  ownerNamespacedName,
+		"objectKind":           objectGVK.Kind,
+		"objectNamespacedName": objectNamespacedName,
+	})
+
+	if metav1.IsControlledBy(object, owner) {
+		objectLogger.Debug("Object has correct ownership. No changes necessary.")
+		return nil
+	}
+
+	actualOwner := metav1.GetControllerOf(object)
+	if actualOwner != nil {
+		// The controller owner doesn't match what the label says should be the owner. Let's correct it!
+		objectLogger.Warn("Object found with ownership label that doesn't match actual owner. Fixing.")
+
+		newOwnerReferences := []metav1.OwnerReference{}
+
+		// Create a list of all of the owner references except for the controller owner reference.
+		for _, existingOwnerReference := range object.GetOwnerReferences() {
+			if existingOwnerReference.Controller != nil && !*existingOwnerReference.Controller {
+				// Don't add the controller owner reference since we're going to reset it later.
+				newOwnerReferences = append(newOwnerReferences, existingOwnerReference)
+			}
+		}
+
+		object.SetOwnerReferences(newOwnerReferences)
+	}
+
+	// The owner reference wasn't found, so we need to add it.
+	if err := controllerutil.SetControllerReference(owner, object, scheme); err != nil {
+		objectLogger.WithError(err).Warn("could not set the controller owner reference")
+		return nil // Not returning error because that could stop the overall reconciliation. Just logging a warning.
+	}
```

Comment:
Can I offer an alternative to the logic in this function that leverages some existing code in library-go?
```
	// Remove any other controller ref
	for i, ref := range object.GetOwnerReferences() {
		if ref.Controller != nil && *ref.Controller {
			if ref.UID != owner.GetUID() {
				ownerRefs := object.GetOwnerReferences()
				ownerRefs[i] = ownerRefs[len(ownerRefs)-1]
				object.SetOwnerReferences(ownerRefs[:len(ownerRefs)-1])
			}
			break
		}
	}

	ownerRef := metav1.NewControllerRef(owner, ownerGVK)
	ownerRefsChanged := librarygocontroller.EnsureOwnerRef(object, *ownerRef)
	
	if ownerRefsChanged {
		objectLogger.Debug("Object has correct ownership. No changes necessary.")
		return nil
	}
```

Reply:
Absolutely! The only reason that I didn't use it because I didn't previously know about it and I didn't find it when looking.

I much prefer using built in api calls. :)

Will Fix.